crates/common/src/shell.rs

```diff
@@ -0,0 +1,260 @@
+        match self {
+            ColorChoice::Always => AnstreamColorChoice::Always,
+            ColorChoice::Never => AnstreamColorChoice::Never,
+            ColorChoice::Auto => AnstreamColorChoice::Auto,
+        }
+    }
+}
+
+/// The output mode: either normal output or completely quiet.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum OutputMode {
+    Normal,
+    Quiet,
+}
+
+/// A shell abstraction that tracks verbosity level, an output mode,
+/// and a color choice.
+/// It uses anstream’s AutoStream for stdout and stderr.
+#[derive(Debug)]
+pub struct Shell {
+    /// Verbosity level (currently unused, but will be in #577)
+    pub verbosity: u8,
+    /// Whether to output anything at all.
+    pub output_mode: OutputMode,
+    /// Whether to use colors.
+    pub color_choice: ColorChoice,
+    /// Adaptive stdout.
+    stdout: AutoStream<std::io::Stdout>,
+    /// Adaptive stderr.
+    stderr: AutoStream<std::io::Stderr>,
+}
+
+impl Shell {
+    /// Creates a new shell with default settings:
+    /// - verbosity 0,
+    /// - Normal output mode,
+    /// - Auto color detection.
+    pub fn new() -> Self {
+        let color = ColorChoice::Auto;
+        Self {
+            verbosity: 0,
+            output_mode: OutputMode::Normal,
+            color_choice: color,
+            stdout: AutoStream::new(std::io::stdout(), color.to_anstream()),
+            stderr: AutoStream::new(std::io::stderr(), color.to_anstream()),
+        }
+    }
+
+    /// Print a string to stdout.
+    pub fn print_out(&mut self, msg: &str) -> io::Result<()> {
+        if self.output_mode == OutputMode::Quiet {
+            return Ok(());
+        }
+        write!(self.stdout, "{}", msg)?;
+        self.stdout.flush()
+    }
+
+    /// Print a line (with a newline) to stdout.
+    pub fn println_out(&mut self, msg: &str) -> io::Result<()> {
+        if self.output_mode == OutputMode::Quiet {
+            return Ok(());
+        }
+        writeln!(self.stdout, "{}", msg)?;
+        self.stdout.flush()
+    }
+
+    /// Print a string to stderr.
+    pub fn print_err(&mut self, msg: &str) -> io::Result<()> {
+        if self.output_mode == OutputMode::Quiet {
+            return Ok(());
+        }
+        write!(self.stderr, "{}", msg)?;
+        self.stderr.flush()
+    }
+
+    /// Print a line (with a newline) to stderr.
+    pub fn println_err(&mut self, msg: &str) -> io::Result<()> {
+        if self.output_mode == OutputMode::Quiet {
+            return Ok(());
+        }
+        writeln!(self.stderr, "{}", msg)?;
+        self.stderr.flush()
+    }
+
+    /// Print a warning message.
+    ///
+    /// If colors are enabled, the “Warning:” prefix is printed in yellow.
+    pub fn warn(&mut self, msg: &str) -> io::Result<()> {
+        let prefix = if self.should_color() {
+            format!("{}Warning:{} ", WARN, Reset)
+        } else {
+            "Warning: ".to_string()
+        };
+        self.println_err(&format!("{}{}", prefix, msg))
```

Comment:
Nit (here and in other places): extra allocations seem unjustified, we could easily just write to stdout/stderr.
Not that it's a performance concern, just feels not very clean. In that regard, I like foundry approach more.